pkg/controller/backup_storage_location_controller.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2020 the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"encoding/json"
+	"strings"
+	"time"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	v1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	velerov1client "github.com/vmware-tanzu/velero/pkg/generated/clientset/versioned/typed/velero/v1"
+	velerov1listers "github.com/vmware-tanzu/velero/pkg/generated/listers/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/persistence"
+	"github.com/vmware-tanzu/velero/pkg/plugin/clientmgmt"
+)
+
+type backupStorageLocationController struct {
+	*genericController
+
+	namespace                   string
+	defaultBackupLocation       string
+	backupLocationClient        velerov1client.BackupStorageLocationsGetter
+	backupStorageLocationLister velerov1listers.BackupStorageLocationLister
+	newPluginManager            func(logrus.FieldLogger) clientmgmt.Manager
+	newBackupStore              func(*velerov1api.BackupStorageLocation, persistence.ObjectStoreGetter, logrus.FieldLogger) (persistence.BackupStore, error)
+}
+
+func NewBackupStorageLocationController(
+	namespace string,
+	defaultBackupLocation string,
+	backupLocationClient velerov1client.BackupStorageLocationsGetter,
+	backupStorageLocationLister velerov1listers.BackupStorageLocationLister,
+	newPluginManager func(logrus.FieldLogger) clientmgmt.Manager,
+	logger logrus.FieldLogger,
+) Interface {
+	c := backupStorageLocationController{
+		genericController:           newGenericController("backup-storage-location", logger),
+		namespace:                   namespace,
+		defaultBackupLocation:       defaultBackupLocation,
+		backupLocationClient:        backupLocationClient,
+		backupStorageLocationLister: backupStorageLocationLister,
+
+		// use variables to refer to these functions so they can be
+		// replaced with fakes for testing.
+		newPluginManager: newPluginManager,
+		newBackupStore:   persistence.NewObjectBackupStore,
+	}
+
+	c.resyncFunc = c.run
+	c.resyncPeriod = 30 * time.Second
```

Comment:
So this will try to revalidate every BSL once every 30 seconds?

We may want to make this tuneable, as there can be monetary access costs associated with hitting services like S3. This change moves the validation from a one-time startup operation to a recurring one, so I don't want to surprise people with new bills :)

I'm also wondering if we want to add a "last validated" timestamp to our BSLs. This may be useful in diagnosing issues, though I'm not sure it can be used to judge when to trigger a validation again. After all, just because it was valid 5 minutes ago doesn't mean it's valid now.

Reply:
Correct.

Yes, good point. I'll add a server flag for this to be configurable. And I like the idea of time-stamping, I think if people make the resync period too long, and with one too many BS:Ls, this will be useful and it's not a big deal to add.

Maybe in the future we can have a command that triggers the validation.

Reply:
What should the default be?